File: script-gen-ui/src/app/model/chan_data/biasChannel.ts
import { IBiasChannel } from '../interface';
import { ParameterFloat } from '../sweep_data/SweepTimingConfig';
import { CommonChanAttributes } from './defaultChannel';

export class BiasChannel {
  common_chan_attributes: CommonChanAttributes;
  bias: ParameterFloat;

  constructor(data: IBiasChannel) {
    this.common_chan_attributes = new CommonChanAttributes(
      data.common_chan_attributes
    );
    this.bias = new ParameterFloat(data.bias);
  }

  toJSON() {
    return {
      common_chan_attributes: this.common_chan_attributes.toJSON(),
      bias: this.bias.toJSON(),
    };
  }
}

File: script-gen-manager/src/model/chan_data/bias_channel.rs
use serde::{Deserialize, Serialize};

use super::default_channel::CommonChanAttributes;
use crate::{
    device::Device, instr_metadata::base_metadata::BaseMetadata,
    model::sweep_data::parameters::ParameterFloat,
};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BiasChannel {
    pub common_chan_attributes: CommonChanAttributes,
    pub bias: ParameterFloat,
}

impl BiasChannel {
    pub fn new(chan_name: String, device: Device) -> Self {
        let mut bias_channel = BiasChannel {
            common_chan_attributes: CommonChanAttributes::new(chan_name, device),
            bias: ParameterFloat::new("bias", 0.0, Some(BaseMetadata::UNIT_VOLTS.to_string())),
        };

        // Call set_defaults on the common_chan_attributes field
        bias_channel.common_chan_attributes.set_defaults();

        bias_channel
    }

    pub fn evaluate(&mut self) {
        self.common_chan_attributes.evaluate();
        self.determine_bias_value();
    }

    fn determine_bias_value(&mut self) {
        if let Some(bias_unit) = &self.bias.unit {
            if bias_unit == &self.common_chan_attributes.source_range.unit {
                self.bias.value = self
                    .common_chan_attributes
                    .source_range
                    .limit(self.bias.value);
            } else {
                //default to 0.0 if source function has been changed
                self.bias.value = self.common_chan_attributes.source_range.limit(0.0);
                self.bias.unit = Some(self.common_chan_attributes.source_range.unit.clone());
            }
        } else {
            //TODO: handle error condition
            println!("bias.unit is None");
        }
    }
}
